File: Samples/java/src/main/java/com/phunware/java/sample/RoutingActivity.java
package com.phunware.java.sample;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.animation.AnimatorSet;
import android.animation.ObjectAnimator;
import android.app.Dialog;
import android.content.Context;
import android.content.DialogInterface;
import android.location.Location;
import android.os.Bundle;
import android.support.design.widget.FloatingActionButton;
import android.support.design.widget.Snackbar;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.CheckBox;
import android.widget.ImageButton;
import android.widget.RelativeLayout;
import android.widget.Spinner;

import com.google.android.gms.maps.CameraUpdate;
import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.model.LatLng;
import com.phunware.core.PwCoreSession;
import com.phunware.core.PwLog;
import com.phunware.location.provider_managed.ManagedProviderFactory;
import com.phunware.location.provider_managed.PwManagedLocationProvider;
import com.phunware.location_core.PwLocationProvider;
import com.phunware.mapping.MapFragment;
import com.phunware.mapping.OnPhunwareMapReadyCallback;
import com.phunware.mapping.PhunwareMap;
import com.phunware.mapping.manager.Callback;
import com.phunware.mapping.manager.Navigator;
import com.phunware.mapping.manager.PhunwareMapManager;
import com.phunware.mapping.manager.Router;
import com.phunware.mapping.model.Building;
import com.phunware.mapping.model.FloorOptions;
import com.phunware.mapping.model.PointOptions;
import com.phunware.mapping.model.RouteManeuverOptions;
import com.phunware.mapping.model.RouteOptions;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.List;

public class RoutingActivity extends AppCompatActivity implements OnPhunwareMapReadyCallback,
        Navigator.OnManeuverChangedListener {
    private static final String TAG = LocationModesActivity.class.getSimpleName();
    private static final int ITEM_ID_LOCATION = -2;

    private PhunwareMapManager mapManager;
    private Building currentBuilding;
    private RelativeLayout content;
    private Spinner floorSpinner;
    private ArrayAdapter<FloorOptions> floorSpinnerAdapter;

    // Navigation Views
    private FloatingActionButton fab;
    private Navigator navigator;
    private View navOverlayContainer;
    private NavigationOverlayView navOverlay;
    private Spinner startPicker;
    private Spinner endPicker;
    private CheckBox accessible;

    View.OnClickListener selectRouteListener = new View.OnClickListener() {
        @Override
        public void onClick(View v) {
            showRoutingDialog();
        }
    };
    View.OnClickListener exitNavListener = new View.OnClickListener() {
        @Override
        public void onClick(View v) {
            stopNavigating();
        }
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.routing);
        content = findViewById(R.id.content);

        // Initialize views for routing
        fab = findViewById(R.id.fab);
        fab.setVisibility(View.GONE);
        fab.setOnClickListener(selectRouteListener);
        navOverlayContainer = findViewById(R.id.nav_overlay_container);
        navOverlay = findViewById(R.id.nav_overlay);

        floorSpinner = findViewById(R.id.floorSpinner);
        floorSpinnerAdapter = new FloorAdapter(this);
        floorSpinner.setAdapter(floorSpinnerAdapter);
        floorSpinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                FloorOptions floor = floorSpinnerAdapter.getItem((int) id);
                if (currentBuilding != null && floor != null) {
                    currentBuilding.selectFloor(floor.getLevel());
                }
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {
            }
        });

        // Register the Phunware API keys
        PwCoreSession.getInstance().setEnvironment(PwCoreSession.Environment.DEV); // FIXME: REMOVE
        PwCoreSession.getInstance().registerKeys(this);

        // Create the map manager and fragment used to load the building
        mapManager = PhunwareMapManager.create(this);
        MapFragment mapFragment = (MapFragment) getFragmentManager().findFragmentById(R.id.map);

        if (mapFragment != null) {
            mapFragment.getPhunwareMapAsync(this);
        }
    }

    @Override
    public void onBackPressed() {
        if (navOverlayContainer.getVisibility() == View.VISIBLE) {
            stopNavigating();
        } else {
            super.onBackPressed();
        }
    }

    @Override
    public void onPhunwareMapReady(final PhunwareMap phunwareMap) {
        // Retrieve buildingId from integers.xml
        int buildingId = getResources().getInteger(R.integer.buildingId);

        mapManager.setPhunwareMap(phunwareMap);
        mapManager.addBuilding(buildingId,
                new Callback<Building>() {
                    @Override
                    public void onSuccess(Building building) {
                        Log.d(TAG, "Building loaded successfully");
                        currentBuilding = building;

                        // Populate floor spinner
                        floorSpinnerAdapter.clear();
                        floorSpinnerAdapter.addAll(building.getBuildingOptions().getFloors());

                        // Initialize a location provider
                        setManagedLocationProvider(building);

                        // Set building to initial floor value
                        FloorOptions initialFloor = building.initialFloor();
                        building.selectFloor(initialFloor.getLevel());

                        // Animate the camera to the building at an appropriate zoom level
                        CameraUpdate cameraUpdate = CameraUpdateFactory
                                .newLatLngBounds(initialFloor.getBounds(), 4);
                        phunwareMap.getGoogleMap().animateCamera(cameraUpdate);

                        // Enabled fab for routing
                        showFab(true);
                    }

                    @Override
                    public void onFailure(Throwable throwable) {
                        Log.d(TAG, "Error when loading building -- " + throwable.getMessage());
                        showFab(false);
                    }
                });
    }

    private void setManagedLocationProvider(Building building) {
        ManagedProviderFactory.ManagedProviderFactoryBuilder builder =
                new ManagedProviderFactory.ManagedProviderFactoryBuilder();
        builder.application(getApplication())
                .context(new WeakReference<Context>(getApplication()))
                .buildingId(String.valueOf(building.getId()));
        ManagedProviderFactory factory = builder.build();
        PwManagedLocationProvider managedProvider
                = (PwManagedLocationProvider) factory.createLocationProvider();
        mapManager.setLocationProvider(managedProvider, building);
        mapManager.setMyLocationEnabled(true);
    }

    private void showFab(final boolean show) {
        final float start = show ? 0 : 1;
        final float end = show ? 1 : 0;
        List<Animator> anims = new ArrayList<>(2);
        anims.add(ObjectAnimator.ofFloat(fab, View.SCALE_X, start, end));
        anims.add(ObjectAnimator.ofFloat(fab, View.SCALE_Y, start, end));
        AnimatorSet s = new AnimatorSet();
        s.playTogether(anims);
        s.addListener(new AnimatorListenerAdapter() {
            @Override
            public void onAnimationStart(Animator animation) {
                if (show) fab.setVisibility(View.VISIBLE);
            }

            @Override
            public void onAnimationEnd(Animator animation) {
                if (!show) fab.setVisibility(View.GONE);
            }
        });
        s.start();
    }

    private void showRoutingDialog() {
        AlertDialog.Builder builder = new AlertDialog.Builder(this);

        // Load custom dialog view
        LayoutInflater inflater = getLayoutInflater();
        View dialogView = inflater.inflate(R.layout.dialog_route_selection, null);
        initDialogUI(dialogView);

        builder.setView(dialogView)
                .setTitle("Select a Route")
                .setMessage("Choose two points to route between")
                .setCancelable(false)
                .setPositiveButton("Route", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int id) {
                        getRoutes();
                    }
                })
                .setNegativeButton("Cancel", new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialog, int id) {
                        // Do Nothing - Close Dialog
                    }
                });

        Dialog d = builder.create();
        d.show();
    }

    private void initDialogUI(View dialogView) {
        startPicker = dialogView.findViewById(R.id.start);
        endPicker = dialogView.findViewById(R.id.end);
        accessible = dialogView.findViewById(R.id.accessible);
        ImageButton reverse = dialogView.findViewById(R.id.reverse);

        reverse.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                onReverseClicked();
            }
        });

        List<PointOptions> points = new ArrayList<>();
        if (currentBuilding != null) {
            for (FloorOptions floorOptions : currentBuilding.getFloorOptions()) {
                if (floorOptions != null && floorOptions.getPoiOptions() != null) {
                    points.addAll(floorOptions.getPoiOptions());
                }
            }
        }

        boolean hasCurrentLocation = false;
        if (mapManager.isMyLocationEnabled() && mapManager.getCurrentLocation() != null) {
            Location myLocation = mapManager.getCurrentLocation();
            LatLng currentLocation = new LatLng(myLocation.getLatitude(), myLocation.getLongitude());
            long currentFloorId = mapManager.getCurrentBuilding().getSelectedFloor().getId();
            if (myLocation.getExtras() != null && myLocation.getExtras()
                    .containsKey(PwLocationProvider.LOCATION_EXTRAS_KEY_FLOOR_ID)) {
                currentFloorId = myLocation.getExtras()
                        .getLong(PwLocationProvider.LOCATION_EXTRAS_KEY_FLOOR_ID);
            }
            points.add(0, new PointOptions()
                    .id(ITEM_ID_LOCATION)
                    .location(currentLocation)
                    .level(currentFloorId)
                    .name(getString(R.string.current_location)));
            hasCurrentLocation = true;
        }

        startPicker.setAdapter(new BuildingAdapter(this, points, getString(R.string.start_prompt)));
        endPicker.setAdapter(new BuildingAdapter(this, points, getString(R.string.end_prompt)));

        if (hasCurrentLocation) {
            startPicker.setSelection(1, false);
        }
    }

    private void getRoutes() {
        long startId = startPicker.getSelectedItemId();
        long endId = endPicker.getSelectedItemId();
        boolean isAccessible = accessible.isChecked();

        Router router = mapManager.findRoutes(startId, endId, isAccessible);
        if (router != null) {
            RouteOptions route = router.shortestRoute();
            if (route == null) {
                PwLog.e(TAG, "Couldn't find route.");
                Snackbar.make(content, R.string.no_route,
                        Snackbar.LENGTH_SHORT).show();
            } else {
                startNavigating(route);
            }
        }
    }

    private void onReverseClicked() {
        int startPos = startPicker.getSelectedItemPosition();
        int endPos = endPicker.getSelectedItemPosition();

        startPicker.setSelection(endPos);
        endPicker.setSelection(startPos);
    }

    private void startNavigating(RouteOptions route) {
        if (navigator != null) {
            navigator.terminate();
        }
        navigator = mapManager.navigate(route);
        navigator.addOnManeuverChangedListener(this);

        navOverlay.setNavigator(navigator);
        navOverlayContainer.setVisibility(View.VISIBLE);
        fab.setImageResource(R.drawable.ic_clear_white);
        fab.setOnClickListener(exitNavListener);

        navigator.start();
    }

    private void stopNavigating() {
        if (navigator != null) {
            navigator.terminate();
            navigator = null;
        }
        navOverlayContainer.setVisibility(View.GONE);
        fab.setImageResource(R.drawable.ic_navigation);
        fab.setOnClickListener(selectRouteListener);
    }

    @Override
    public void onManeuverChanged(Navigator navigator, int position) {
        // Update the selected floor when the maneuver floor changes
        RouteManeuverOptions maneuver = navigator.getManeuvers().get(position);
        int selectedPosition = floorSpinner.getSelectedItemPosition();
        for (int i = 0; i < floorSpinnerAdapter.getCount(); i++) {
            FloorOptions floor = floorSpinnerAdapter.getItem(i);
            if (selectedPosition != i && floor != null && floor.getId() == maneuver.getFloorId()) {
                floorSpinner.setSelection(i);
            }
        }
    }

    @Override
    public void onRouteSnapFailed() {
        // Do Nothing
    }
}

File: Samples/kotlin/src/main/java/com/phunware/kotlin/sample/RoutingActivity.kt
package com.phunware.kotlin.sample

/* Copyright (C) 2018 Phunware, Inc.

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
"Software"), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL Phunware, Inc. BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

Except as contained in this notice, the name of Phunware, Inc. shall
not be used in advertising or otherwise to promote the sale, use or
other dealings in this Software without prior written authorization
from Phunware, Inc. */

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.animation.AnimatorSet
import android.animation.ObjectAnimator
import android.os.Bundle
import android.support.design.widget.FloatingActionButton
import android.support.design.widget.Snackbar
import android.support.v7.app.AlertDialog
import android.support.v7.app.AppCompatActivity
import android.util.Log
import android.view.View
import android.widget.*
import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.MapStyleOptions
import com.phunware.core.PwCoreSession
import com.phunware.core.PwLog
import com.phunware.location.provider_managed.ManagedProviderFactory
import com.phunware.location.provider_managed.PwManagedLocationProvider
import com.phunware.location_core.PwLocationProvider
import com.phunware.mapping.MapFragment
import com.phunware.mapping.OnPhunwareMapReadyCallback
import com.phunware.mapping.PhunwareMap
import com.phunware.mapping.manager.Callback
import com.phunware.mapping.manager.Navigator
import com.phunware.mapping.manager.PhunwareMapManager
import com.phunware.mapping.manager.Router
import com.phunware.mapping.model.Building
import com.phunware.mapping.model.FloorOptions
import com.phunware.mapping.model.PointOptions
import com.phunware.mapping.model.RouteOptions
import java.lang.ref.WeakReference
import java.util.*

class RoutingActivity : AppCompatActivity(), OnPhunwareMapReadyCallback,
        Navigator.OnManeuverChangedListener, Building.OnFloorChangedListener {

    private lateinit var mapManager: PhunwareMapManager
    private lateinit var currentBuilding: Building
    private lateinit var content: RelativeLayout
    private lateinit var floorSpinner: Spinner
    private lateinit var floorSpinnerAdapter: ArrayAdapter<FloorOptions>

    // Navigation Views
    private lateinit var fab: FloatingActionButton
    private var navigator: Navigator? = null
    private lateinit var navOverlayContainer: View
    private lateinit var navOverlay: NavigationOverlayView
    private lateinit var startPicker: Spinner
    private lateinit var endPicker: Spinner
    private lateinit var accessible: CheckBox
    private lateinit var currentLocationPoint : PointOptions

    private var selectRouteListener: View.OnClickListener = View.OnClickListener { showRoutingDialog() }
    private var exitNavListener: View.OnClickListener = View.OnClickListener { stopNavigating() }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.routing)
        content = findViewById(R.id.content)

        // Initialize views for routing
        fab = findViewById(R.id.fab)
        fab.hide()
        fab.setOnClickListener(selectRouteListener)
        navOverlayContainer = findViewById(R.id.nav_overlay_container)
        navOverlay = findViewById(R.id.nav_overlay)

        floorSpinner = findViewById(R.id.floorSpinner)
        floorSpinnerAdapter = FloorAdapter(this)
        floorSpinner.adapter = floorSpinnerAdapter
        floorSpinner.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(parent: AdapterView<*>, view: View, position: Int, id: Long) {
                val floor = floorSpinnerAdapter.getItem(id.toInt())
                if (floor != null) {
                    currentBuilding.selectFloor(floor.level)
                }
            }

            override fun onNothingSelected(parent: AdapterView<*>) {}
        }

        // Register the Phunware API keys
        PwCoreSession.getInstance().registerKeys(this)

        // Create the map manager and fragment used to load the building
        mapManager = PhunwareMapManager.create(this)

        if (savedInstanceState == null) {
            fragmentManager.beginTransaction()
                .replace(R.id.map, CustomMapFragment.newInstance(this, this))
                .commit()
        }
    }

    override fun onBackPressed() {
        if (navOverlayContainer.visibility == View.VISIBLE) {
            stopNavigating()
        } else {
            super.onBackPressed()
        }
    }

    override fun onPhunwareMapReady(phunwareMap: PhunwareMap) {
        // Retrieve buildingId from integers.xml
        val buildingId = resources.getInteger(R.integer.buildingId)

        phunwareMap.googleMap.uiSettings.isMapToolbarEnabled = false
        phunwareMap.googleMap.setMapStyle(MapStyleOptions.loadRawResourceStyle(
                this@RoutingActivity, R.raw.map_style))

        mapManager.setPhunwareMap(phunwareMap)
        mapManager.addBuilding(buildingId.toLong(),
                object : Callback<Building> {
                    override fun onSuccess(building: Building) {
                        Log.d(TAG, "Building loaded successfully")
                        currentBuilding = building

                        // Populate floor spinner
                        floorSpinnerAdapter.clear()
                        floorSpinnerAdapter.addAll(building.buildingOptions.floors)

                        // Add a listener to monitor floor switches
                        mapManager.addFloorChangedListener(this@RoutingActivity)

                        // Initialize a location provider
                        setManagedLocationProvider(building)

                        // Set building to initial floor value
                        val initialFloor = building.initialFloor()
                        building.selectFloor(initialFloor.level)

                        // Animate the camera to the building at an appropriate zoom level
                        val cameraUpdate = CameraUpdateFactory
                                .newLatLngBounds(initialFloor.bounds, 4)
                        phunwareMap.googleMap.animateCamera(cameraUpdate)

                        // Enabled fab for routing
                        showFab(true)
                    }

                    override fun onFailure(throwable: Throwable) {
                        Log.d(TAG, "Error when loading building -- " + throwable.message)
                        showFab(false)
                    }
                })
    }

    private fun setManagedLocationProvider(building: Building) {
        val builder = ManagedProviderFactory.ManagedProviderFactoryBuilder()
        builder.application(application)
                .context(WeakReference(application))
                .buildingId(building.id.toString())
        val factory = builder.build()
        val managedProvider = factory.createLocationProvider() as PwManagedLocationProvider
        mapManager.setLocationProvider(managedProvider, building)
        mapManager.isMyLocationEnabled = true
    }

    private fun showFab(show: Boolean) {
        runOnUiThread {
            val start = (if (show) 0 else 1).toFloat()
            val end = (if (show) 1 else 0).toFloat()
            val anims = ArrayList<Animator>(2)
            anims.add(ObjectAnimator.ofFloat(fab, View.SCALE_X, start, end))
            anims.add(ObjectAnimator.ofFloat(fab, View.SCALE_Y, start, end))
            val s = AnimatorSet()
            s.playTogether(anims)
            s.addListener(object : AnimatorListenerAdapter() {
                override fun onAnimationStart(animation: Animator) {
                    if (show) fab.show()
                }

                override fun onAnimationEnd(animation: Animator) {
                    if (!show) fab.hide()
                }
            })
            s.start()
        }
    }

    private fun showRoutingDialog() {
        val builder = AlertDialog.Builder(this)

        // Load custom dialog view
        val inflater = layoutInflater
        val dialogView = inflater.inflate(R.layout.dialog_route_selection, null)
        initDialogUI(dialogView)

        builder.setView(dialogView)
                .setTitle("Select a Route")
                .setMessage("Choose two points to route between")
                .setCancelable(false)
                .setPositiveButton("Route", { _, _ -> getRoutes() })
                .setNegativeButton("Cancel", { _, _ ->
                    // Do Nothing - Close Dialog
                })

        val d = builder.create()
        d.show()
    }

    private fun initDialogUI(dialogView: View) {
        startPicker = dialogView.findViewById(R.id.start)
        endPicker = dialogView.findViewById(R.id.end)
        accessible = dialogView.findViewById(R.id.accessible)
        val reverse = dialogView.findViewById<ImageButton>(R.id.reverse)

        reverse.setOnClickListener { onReverseClicked() }

        val points = ArrayList<PointOptions>()

        currentBuilding.floorOptions
                .filter { it != null && it.poiOptions != null }
                .forEach { points.addAll(it.poiOptions) }


        var hasCurrentLocation = false
        if (mapManager.isMyLocationEnabled && mapManager.currentLocation != null) {
            val myLocation = mapManager.currentLocation
            val currentLocation = LatLng(myLocation.latitude, myLocation.longitude)
            var currentFloorId = mapManager.currentBuilding.selectedFloor.id
            if (myLocation.extras != null && myLocation.extras
                            .containsKey(PwLocationProvider.LOCATION_EXTRAS_KEY_FLOOR_ID)) {
                currentFloorId = myLocation.extras
                        .getLong(PwLocationProvider.LOCATION_EXTRAS_KEY_FLOOR_ID)
            }

            currentLocationPoint = PointOptions()
                    .id(ITEM_ID_LOCATION)
                    .location(currentLocation)
                    .floorId(currentFloorId)
                    .name(getString(R.string.current_location))
            points.add(0, currentLocationPoint)
            hasCurrentLocation = true
        }

        startPicker.adapter = BuildingAdapter(this, points, getString(R.string.start_prompt))
        endPicker.adapter = BuildingAdapter(this, points, getString(R.string.end_prompt))

        if (hasCurrentLocation) {
            startPicker.setSelection(1, false)
        }
    }

    private fun getRoutes() {
        val startId = startPicker.selectedItemId
        val endId = endPicker.selectedItemId
        val isAccessible = accessible.isChecked

        val router = if (startId == ITEM_ID_LOCATION) {
            mapManager.findRoutes(currentLocationPoint.location, endId,
                    currentLocationPoint.floorId, isAccessible)
        } else if (endId == ITEM_ID_LOCATION) {
            mapManager.findRoutes(startId, currentLocationPoint.location,
                    currentLocationPoint.floorId, isAccessible)
        } else {
            mapManager.findRoutes(startId, endId, isAccessible)
        }

        if (router != null) {
            val route = router.shortestRoute()
            if (route == null) {
                PwLog.e(TAG, "Couldn't find route.")
                Snackbar.make(content, R.string.no_route,
                        Snackbar.LENGTH_SHORT).show()
            } else {
                startNavigating(route)
            }
        }
    }

    private fun onReverseClicked() {
        val startPos = startPicker.selectedItemPosition
        val endPos = endPicker.selectedItemPosition

        startPicker.setSelection(endPos)
        endPicker.setSelection(startPos)
    }

    private fun startNavigating(route: RouteOptions) {
        if (navigator != null) {
            navigator!!.stop()
        }
        navigator = mapManager.navigate(route)
        navigator!!.addOnManeuverChangedListener(this)

        navOverlay.setNavigator(navigator!!)
        navOverlayContainer.visibility = View.VISIBLE
        fab.setImageResource(R.drawable.ic_clear_white)
        fab.setOnClickListener(exitNavListener)

        navigator!!.start()
    }

    private fun stopNavigating() {
        if (navigator != null) {
            navigator!!.stop()
            navigator = null
        }
        navOverlayContainer.visibility = View.GONE
        fab.setImageResource(R.drawable.ic_navigation)
        fab.setOnClickListener(selectRouteListener)
    }

    override fun onManeuverChanged(navigator: Navigator, position: Int) {
        // Update the selected floor when the maneuver floor changes
        val maneuver = navigator.maneuvers[position]
        val selectedPosition = floorSpinner.selectedItemPosition
        for (i in 0 until floorSpinnerAdapter.count) {
            val floor = floorSpinnerAdapter.getItem(i)
            if (selectedPosition != i && floor != null && floor.id == maneuver.floorId) {
                floorSpinner.setSelection(i)
            }
        }
    }

    override fun onRouteSnapFailed() {
        // Do Nothing
    }

    override fun onFloorChanged(building: Building?, floorId: Long) {
        for (index in 0 until floorSpinnerAdapter.count) {
            val floor = floorSpinnerAdapter.getItem(index)
            if (floor != null && floor.id == floorId) {
                if (floorSpinner.selectedItemPosition != index) {
                    runOnUiThread { floorSpinner.setSelection(index) }
                    break
                }
            }
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        mapManager?.let {
            it.onDestroy()
        }
    }

    companion object {
        private val TAG = LocationModesActivity::class.java.simpleName
        private val ITEM_ID_LOCATION = -2L
    }
}
